pp_hot.c

```diff
@@ -5177,16 +5177,12 @@ PP(pp_entersub)
 
         sv = *svp;
 
-        /* TODO: This is currently horribly inefficient. It can likely be made
-         * a lot better by adjusting the code lower down instead
-         * For now we'll just move all the stack args down one position
-         */
-        PERL_UNUSED_RESULT(POPs);
+        /* Move all the stack args down one position to where everyone expects
+         * them */
+        Size_t count = SP - mark;
+        Move(svp+1, svp, count, SV *);
 
-        while(svp <= SP) {
-            svp[0] = svp[1];
-            svp++;
-        }
+        PERL_UNUSED_RESULT(POPs);
```

Comment:
It is slightly more efficient, but the old version compiled down to memmove() too (details below).

I do wonder if changing this is worth the performance cost, 

Tested with:
```
perf stat ./miniperl -e 'my $foo = sub  { }; my @args = ( 1 .. 10 ); for (1..1_000_000) { $foo->(@args) }'
```

After 3 trials and taking the minimum of each set (perf includes interrupt processing) this appears to make the test run about 4% slower in instructions and user time.

<details><summary>Raw performance numbers</summary>

Before the patch:
```
       225,616,164      cycles                    #    4.529 GHz                    
       777,104,152      instructions              #    3.44  insn per cycle         
       155,617,016      branches                  # 3123.946 M/sec                  
            20,608      branch-misses             #    0.01% of all branches    
       0.050176000 seconds user

       247,633,809      cycles                    #    4.360 GHz                    
       777,100,981      instructions              #    3.14  insn per cycle         
       155,617,656      branches                  # 2739.773 M/sec                  
            20,560      branch-misses             #    0.01% of all branches    
       0.057839000 seconds user

       230,714,487      cycles                    #    4.556 GHz                    
       777,107,478      instructions              #    3.37  insn per cycle         
       155,618,909      branches                  # 3073.204 M/sec                  
            20,558      branch-misses             #    0.01% of all branches    
       0.050940000 seconds user
```
With the patch:
```
       245,858,270      cycles                    #    4.591 GHz                    
       812,074,404      instructions              #    3.30  insn per cycle         
       165,612,897      branches                  # 3092.589 M/sec                  
       20,359      branch-misses             #    0.01% of all branches
       0.053863000 seconds user

       238,782,598      cycles                    #    4.598 GHz                    
       812,090,195      instructions              #    3.40  insn per cycle         
       165,614,388      branches                  # 3188.989 M/sec                  
            20,394      branch-misses             #    0.01% of all branches    
       0.052243000 seconds user

       243,833,810      cycles                    #    4.404 GHz                    
       812,133,770      instructions              #    3.33  insn per cycle         
       165,624,467      branches                  # 2991.232 M/sec                  
            20,408      branch-misses             #    0.01% of all branches    
       0.056446000 seconds user
```

</details>

<details><summary>Generated code for this block</summary>

Before:
```
L3123:
        movq    PL_markstack_ptr(%rip), %rdx
        movslq  (%rdx), %rcx
        movq    PL_stack_base(%rip), %rdx
        leaq    (%rdx,%rcx,8), %rsi
        movq    8(%rsi), %rbp
        leaq    8(%rsi), %rdi
        movq    %rbp, 16(%rsp)
        cmpq    %rdi, %r13
        jb      .L2943
        subq    $16, %rax
        subq    %rsi, %rax
        addq    $16, %rsi
        shrq    $3, %rax
        leaq    8(,%rax,8), %rdx
        call    memmove@PLT
        jmp     .L2943
```

After:

```
.L3123:
        movq    PL_markstack_ptr(%rip), %rax
        movslq  (%rax), %rdx
        movq    PL_stack_base(%rip), %rax
        leaq    (%rax,%rdx,8), %rsi
        movq    %rbx, %rdx
        movq    8(%rsi), %rbp
        leaq    8(%rsi), %rdi
        movq    %rbp, 16(%rsp)
        subq    %rsi, %rdx
        js      .L3000
        addq    $16, %rsi
        leaq    -8(%rbx), %r13
        call    memmove@PLT
        jmp     .L2944
```

</details>

Reply:
> I do wonder if changing this is worth the performance cost,
> ...
> After 3 trials and taking the minimum of each set (perf includes interrupt processing) this appears to make the test run about 4% slower in instructions and user time.

Hrm. Indeed - that sounds like a bit of a hit to be taking here :(